### PR TITLE
Run Prometheus exporter inside IC Binary

### PR DESCRIPTION
**Pull request by Rulox**
### Proposed changes
Instead of using the NGINX Prometheus exporter in a container inside IC pod, use it as a library and expose NGINX and NGINX Plus metrics in a Prometheus format from the same binary as the IC is running.

This PR Also updates protobuf library needed to work well with the Prometheus client for Golang

### Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

File: deployments/daemon-set/nginx-ingress.yaml [hunk 1]
```diff
@@ -1,42 +0,0 @@
-apiVersion: extensions/v1beta1
```
Comment (pleshakov) [inline deployments/daemon-set/nginx-ingress.yaml]: this file is allright

File: deployments/daemon-set/nginx-plus-ingress.yaml [hunk 2]
```diff
@@ -1,43 +0,0 @@
-apiVersion: extensions/v1beta1
```
Comment (pleshakov) [inline deployments/daemon-set/nginx-plus-ingress.yaml]: man, you got carried away :)

File: deployments/daemon-set/nginx-ingress.yaml (hunk 1 above)
Comment (pleshakov) [reply deployments/daemon-set/nginx-ingress.yaml]: please make the changes here similar to the changes to the deployment file

File: deployments/daemon-set/nginx-plus-ingress.yaml (hunk 2 above)
Comment (pleshakov) [reply deployments/daemon-set/nginx-plus-ingress.yaml]: please make the changes here similar to the changes to the deployment file

File: deployments/helm-chart/values.yaml [hunk 3]
```diff
@@ -160,19 +160,8 @@ rbac:
   create: true
 
 prometheus:
-  ## Deploys a Prometheus exporter container within the Ingress controller pod. Requires NGINX status enabled via controller.nginxStatus.enable.
-  ## Note: the exporter will use the port specified by controller.nginxStatus.port.
+  ## Runs a Prometheus exporter listener within the Ingress controller binary.
```
Comment (pleshakov) [inline deployments/helm-chart/values.yaml]: the comment here must be the same as the corresponding entry in the doc

File: deployments/daemon-set/nginx-ingress-with-prometheus.yaml [hunk 4]
```diff
@@ -1,54 +0,0 @@
...
-        args:
-          - -nginx-configmaps=$(POD_NAMESPACE)/nginx-config
-          - -default-server-tls-secret=$(POD_NAMESPACE)/default-server-secret
-         #- -v=3 # Enables extensive logging. Useful for troubleshooting.
-         #- -report-ingress-status
-         #- -external-service=nginx-ingress
-         #- -enable-leader-election
-      - image: nginx/nginx-prometheus-exporter:0.2.0
-        name: nginx-prometheus-exporter
-        ports:
```
Comment (pleshakov) [inline deployments/daemon-set/nginx-ingress-with-prometheus.yaml]: could you add this port to the list of the ports in the manifests and helm templates?

File: deployments/daemon-set/nginx-plus-ingress.yaml (hunk 2 above)
Comment (Rulox) [reply deployments/daemon-set/nginx-plus-ingress.yaml]: Oops, yeah I probably missclicked, recovering these

File: deployments/daemon-set/nginx-plus-ingress.yaml [hunk 5]
```diff
@@ -11,6 +11,9 @@ spec:
     metadata:
       labels:
         app: nginx-ingress
+     #annotations:
```
Comment (Dean-Coakley) [inline deployments/daemon-set/nginx-plus-ingress.yaml]: I'd prefer to see a space between `#` and comment lines in yaml. This allows use of the ctrl+/ shortcut on text editors to comment/uncomment yaml config without breaking indentation levels.

File: deployments/helm-chart/templates/controller-daemonset.yaml [hunk 6]
```diff
@@ -84,23 +84,6 @@ spec:
...
-          - -nginx.plus
-          - -nginx.scrape-uri
-          - http://127.0.0.1:{{ .Values.controller.nginxStatus.port }}/api
-{{ else }}
-          - -nginx.scrape-uri
-          - http://127.0.0.1:{{ .Values.controller.nginxStatus.port }}/stub_status
-{{- end }}
-{{- end }}
-{{- end }}
+          - -enable-prometheus-metrics={{ .Values.prometheus.create }}
```
Comment (pleshakov) [inline deployments/helm-chart/templates/controller-daemonset.yaml]: don't forget about the prometheus port in the ports section

File: deployments/helm-chart/templates/controller-deployment.yaml [hunk 7]
```diff
@@ -82,23 +82,6 @@ spec:
...
-          - -nginx.plus
-          - -nginx.scrape-uri
-          - http://127.0.0.1:{{ .Values.controller.nginxStatus.port }}/api
-{{ else }}
-          - -nginx.scrape-uri
-          - http://127.0.0.1:{{ .Values.controller.nginxStatus.port }}/stub_status
-{{- end }}
-{{- end }}
-{{- end }}
+          - -enable-prometheus-metrics={{ .Values.prometheus.create }}
```
Comment (pleshakov) [inline deployments/helm-chart/templates/controller-deployment.yaml]: don't forget about the prometheus port in the ports section

File: docs/installation.md [hunk 8]
```diff
@@ -165,26 +165,14 @@ For NGINX Plus, you can access the live activity monitoring dashboard:
...
-* For a deployment, run:
-    ```
-    $ kubectl apply -f deployment/nginx-plus-ingress-with-prometheus.yaml
-    ```
-* For a daemon set, run:
-    ```
-    $ kubectl apply -f daemon-set/nginx-plus-ingress-with-prometheus.yaml
+You can expose NGINX or NGINX Plus metrics for collection by [Prometheus](https://prometheus.io/):
+
+1. Run the Ingress controller with the `-enable-prometheus-metrics` [command-line argument](cli-arguments.md). As a result, the Ingress Controller will expose NGINX or NGINX Plus metrics in the Prometheus format via the path `/metrics` on port `9113` (customizable via the `-prometheus-metrics-listen-port` command-line argument).
```
Comment (pleshakov) [inline docs/installation.md]: Please add an additional step (2):

1. Add the Prometheus port to the list of the ports of the Ingress Controller container:
```
        - name: prometheus
          containerPort: 9113
```

File: docs/installation.md [hunk 9]
```diff
@@ -165,26 +165,14 @@ For NGINX Plus, you can access the live activity monitoring dashboard:
...
+2. Make the Prometheus aware of the Ingress Controller targets by adding the following annotations to the template of the Ingress Controller pod:
```
Comment (pleshakov) [inline docs/installation.md]:
```suggestion
1. Make the Prometheus aware of the Ingress Controller targets by adding the following annotations to the template of the Ingress Controller pod (note: this assumes your Prometheus is configured to discover targets by analyzing the annotations of pods):
```

File: deployments/daemon-set/nginx-plus-ingress.yaml (hunk 5 above)
Comment (pleshakov) [reply deployments/daemon-set/nginx-plus-ingress.yaml]: @Dean-Coakley  good observation. if anything, that would be a separate task

Comment (Rulox) [reply deployments/daemon-set/nginx-plus-ingress.yaml]: I will keep like this for consistency, maybe we can change all of them in the future if that seems better. I am open to both ways (agree that maybe a space after the `#` is easier to read) thanks!

**Pull request merged**